### ISSUE TITLE
Adding pydl as a dependency for PYPIT

### ISSUE THREAD

**Issue by jhennawi** [pypeit/PypeIt#363]
Hi All, 

We need to have a discussion about adding another dependency to pydl. The package in question is
pydl, which contains ports of codes to python from idlspec2d and idlutils that the SDSS pipelines depended on. Many of these codes are required for my extraction (which now works very well). 

You can look at the code base here:  https://github.com/weaverba137/pydl

@profxj is of the opinion that we should just add these codes to the PYPIT distribution. 

I disagree. Although @pydl does not have a ton of users or developers at the moment, I think we will continue to need to pull some codes from this project as we continue to develop (I know of several codes I already need). 

I don't think this is a dependency will be problematic, since the code is not changing that rapidly. If we find it is causing problems we can go with @profxj approach of simply copying in the codes. 

My main reason for wanting to make it a dependency is that I think that there may be some people using these codes, and we could benefit from their bug fixes. I certainly found a lot and fixed a lot of things. It is also nicely organized with code tests etc., so we could be benefiting from that structure. 

PYPIT already has some complex dependencies, the most problematic is of course linetools which causes all sorts of install issues. I would argue that linetools is worth that headache with the backend etc. But I do not agree that adding pydl is causing lots more work for our users. Basically you just have to git clone it and install. 

Joe

**Comment by profxj** [comment]
Here are my thoughts:

Pros:

- Any improvements we make will benefit the PYDL user-base which I currently estimate as 3 persons and likely to decline.  I doubt the young generation will be keen on touching anything with the stink of IDL.
- The package is an Astropy affiliated package although it is pending removal unless the docs are signficantly improved.  No evidence that has been addressed.

Cons:

- No hot fix capability by any of the PYPIT core developers.  I would insist one of us gain write privilege before adding this as dependency.
- Code is known to have significant bugs, unfinished methods, etc.
- Documentation is dodgy at best
- If @jhennawi struggles to install *linetools*, we should be wary of adding any additional dependencies
- There is only one contributor (Weaver, who is a great coder) and I expect he will not be doing hardly anything additional given it is not used by DESI or any NOAO projects (to my knowledge).  Indeed, he advised us to use the DESI tools.
- Test suite is partial

An examination of the code-frequency chart indicates very little development since 2016:

https://github.com/weaverba137/pydl/graphs/code-frequency

**Comment by jhennawi** [comment]
I don't think the situation is being represented very accurately here. 

1) First, here is what Weaver said when I asked him about pydl: 

"PyDL (https://github.com/weaverba137/pydl) is actively supported, but most of the idlspec2d functionality is geared toward supporting the combine1fiber() function.  Honestly, at this point, I would take a look at the DESI Python code if you need a full Python-based spectroscopic data reduction pipeline.  https://github.com/desihub/desispec and other packages in the same group."

I don't think it is fair to say that he advised us to use the DESI tools. He did suggest we look at the DESI pipeline, but I've done that and the tools we need are not there because spectro-perfectionism does not require things like b-splines. We have integrated DESI ports where they are available, and continue to do so. 

2)  The number of pydl users is being misrepresented. Let me remind you that pydl is an astropy affiliated package, which is more than we can currently say for PYPIT. You  can check this page: 

http://lira.no-ip.org:8080/doc/python-pydl-doc/html/pydl/index.html

I don't think @profxj has a way of estimating that the number of users using these astropy codes is three persons, but I could be wrong.

3)  The documentation at the link above looks a hell of a lot better and up to date than ours. So if that is dodgy at best, not sure what adjective to use for ours. 

4) I don't think it is productive to make statements about "the stink of IDL", or to suggest that "spending time learning how to do that within pypit instead of porting IDLUTILS" is a much better strategy. Here is why: 

a)  Much of PYPIT is already based on ports of IDLUTILS, they just don't live in pydl.  For example algorithms like trace_crude, trace_fweight, trace_gweight, extract_boxcar, extract_asymbox2,  are all idlutils codes that we have ported ourselves into PYPIT. Although they don't live in pydl, arguably these ports belong in pydl since they are ports of idlutils and idlspec2d.  If we were doing that work in pydl in the context of an astropy affiliated package, we might have more people checking and using our codes, because it is astropy affiliated. 

b) There is no evidence supporting the idea that the algorithms that we have developed "natively in python or pypit" (whatever that means) are in any way superior to those that were in IDL. Indeed, the native python bspline we were using was a piece of shit, and has just been rescued by a better b-spline from IDL that I ported  My recent experience also suggests that robust_polyfit has issues that for example the same codes in idlutils did not. I challenge you to name a core routine that we have developed from scratch or natively in python that works better than previous things that existed in idlutils, idlspec2d, or xidl? The only example I can think of is wavelength solutions (i.e. arclines and the holy grail). 

c) It is clear that python is the way forward and IDL is a thing of a past. But in my opinion, it is important not to forget that the algorithms and codes in idlspec2d and idlutils were battle tested by nearly two decades of use in the SDSS data reduction pipelines, and by large number of developers (myself, Prochaska, George Becker, Scott Burles, Schlegel, Rob Simcoe, Rebecca Bernstein) that wrote pipelines based on these routines, the large number of instruments and telescopes that used them, and the larger number of users using those pipelines.  Starting from scratch means we have to do all that debugging over again, and is in my opinion  not an option, although this appears to be the path PYPIT developers are most fond of traveling.

In conclusion, my suggestions are: 

-- We consult with Weaver, and get write access for our core developers as a condition for moving forward. 

-- We consider moving our ports of if idlutils and idlspec2d codes into pydl instead of in PYPIT, since that would allow us to immediately take advantage of their astropy affiliation status, and possibly get more users touching, commenting on, and using our codes.  

-- We live with the dependency unless it is causing headaches that would lead us to reconsider. 

-- We spend more time carefully porting algorithms that are well tested and known to work well, and less time reinvening the wheel by starting from scratch. 

Joe

**Comment by kbwestfall** [comment]
This may already be resolved.  Setting to v0.9 milestone assuming we can resolve it by then.

**Comment by jhennawi** [comment]
I think this proposal may be dead. See Weaver's reply at: 

https://github.com/weaverba137/pydl/pull/43

Joe

**Comment by weaverba137** [comment]
Wow, I really, really wish I'd seen this prior to right now. I feel like I've been deliberately excluded from this conversation.

**Comment by profxj** [comment]
Hard to deliberately exclude anyone from a public GitHub page, @weaverba137

**Comment by weaverba137** [comment]
If this issue had been referenced in the pydl PR, I would have understood the context, and my responses may have been different.

**Comment by profxj** [comment]
Good point.  Am not sure why @jhennawi didn't mention it.

**Comment by weaverba137** [comment]
I'm still open to having a conversation about this.  I didn't realize anyone besides Joe was involved.

**Comment by jhennawi** [comment]
Hi @weaverba137, my intent was not exclude you from the discussion and I apologize for not including you in this thread.  I'm happy to have further discussions about this and have reopened the thread. 

I continue to believe the best way forward would be for us to merge our fixes and enhancements to the pydl codes back into pydl. The changes I've made are fairly significant though, so you need to weigh this carefully. 

You did an amazing job of porting the codes, and your programming is impeccable. That said, as is normal for not heavily tested new codes there were a lot of bugs that I discovered in the bspline when trying to do more sophisticated things like exploit the full breakpoint functionality, and
the polynomial fitting features. Also, your code was not iterating. There are other bugs we fixed when we discovered them throughout. Other changes relate to edge cases where the codes fail when they should not that were also inherent as bugs in the original IDL codes. Those have also been fixed. My changes are all commented by my initials in the vast majority of cases. 

These codes are now a major dependency for the PypeIt project and as such we are not able to compromise on these fixes and changes.  So a merge would need to adopt the default assumption that the PypeIt versions replace pydl versions unless an obvious bug has been discovered. I also think that one of the PypeIt core developers needs to have write access to the main repository, as I'd prefer not deal with syncing forks. 

I'm happy to compromise on coding style and will accept any changes that don't compromise functionality. As I'm not the most experienced python coder, and I made some of these modifications when I was quite green. That said, the bugs I found are real and independent of coding style. 

Joe

**Comment by weaverba137** [comment]
Thanks for the further details.  I'm concerned though that in the conversation on weaverba137/pydl#43, we seemed to be talking past each other.  Let me be perfectly clear: I accept that the bugs you found exist, and that they need to be fixed.  In fact, I have already started to fix some of the simpler problems on another branch.

However, I did not see any evidence that you were addressing my concerns about the proposed changes breaking unit tests, about comparing the changes to the IDL code to satisfy the reproducibility requirement, or even a simple request to add a description of the changes to the change log file.  

Finally, when you asked about granting write access to the repository, I did not have the context (this issue!) to realize that this group may need rapid deployment of bug fixes.  That does change the consideration.  Even then though, it is perfectly reasonable to submit pull requests via a fork.  When you say:

> I'd prefer not deal with syncing forks.

Are you truly speaking for your entire collaboration? Changes need to be submitted via a PR anyway, so it makes no difference whether the branch is in the "main" repository or in a fork.  Even when doing rapid development for DESI, we still submit changes via PRs for careful review, and verify that unit tests are not broken by the change.  If I am misreading the level of urgency given the discussion above, please clarify!

**Comment by weaverba137** [comment]
Hello y'all,  I'm participating in the Astropy Coordination meeting this week, and I plan to do a lot of work on PyDL.  Have you had any further discussions about this?

**Comment by jhennawi** [comment]
We have made so many changes and bug fixes to pydl that I think a merger would required a tremendous concerted effort. We are going to just keep some of the core routines in PypeIt for now until we have more manpower to merge.
